File: Exercism/javascript/circular-buffer/circular-buffer.js
'use strict'


class BufferEmptyError extends Error {
  constructor () {
    super('Buffer empty')
  }
}

class BufferFullError extends Error {
  constructor () {
    super('Buffer full!')
  }
}

class CircularBuffer {
  constructor (int) {
    this.buffer = Array(int).fill(null)
    this.r = 0
    this.w = 0
  }

  get size () {
    return this.buffer.length
  }

  read () {
    const { buffer, r, size } = this
    if (buffer[r] === null) {
      throw new BufferEmptyError()
    }
    const value = buffer[r]
    buffer[r] = null
    this.r = inc(r, size)
    return value
  }

  write (value) {
    if (!isValue(value)) {
      return
    }
    const { buffer, w, size } = this
    if (buffer[w] !== null) {
      throw new BufferFullError()
    }
    buffer[w] = value
    this.w = inc(w, size)
  }

  forceWrite (value) {
    if (!isValue(value)) {
      return
    }
    const { buffer, w, r, size } = this
    buffer[w] = value
    if (r === w) {
      this.r = inc(r, size)
    }
    this.w = inc(w, size)
  }
  
  clear () {
    this.buffer = this.buffer.map(() => null)
  }
}

const inc = (nr, mod) => (nr + 1) % mod
const isValue = (value) =>
  typeof value !== 'undefined' && value !== null

const circularBuffer = (int) => new CircularBuffer(int)

export {
  BufferFullError,
  BufferEmptyError
}

export default circularBuffer

File: Exercism/python/circular-buffer/circular_buffer.py
class BufferFullException(Exception):
    def __init__(self, message='Buffer full!'):
        super().__init__(message)


class BufferEmptyException(Exception):
    def __init__(self, message='Buffer empty!'):
        super().__init__(message)


class CircularBuffer(object):
    """Creates a circular buffer of fixed size"""

    def __init__(self, capacity: int):
        self._rc = 0
        self._wc = 0
        self.capacity = capacity
        self.clear()

    def inc_rc(self) -> None:
        self._rc += 1
        self._rc %= self.capacity

    def inc_wc(self) -> None:
        self._wc += 1
        self._wc %= self.capacity

    def read(self):
        value = self.buffer[self._rc]
        if value is None:
            raise BufferEmptyException
        self.buffer[self._rc] = None
        self.inc_rc()
        return value

    def write(self, data):
        if self.buffer[self._wc] is not None:
            raise BufferFullException
        self.buffer[self._wc] = data
        self.inc_wc()

    def overwrite(self, data):
        if self._wc == self._rc:
            self.inc_rc()
        self.buffer[self._wc] = data
        self.inc_wc()

    def clear(self):
        self.buffer = [None] * self.capacity
